Rebuilt Goal Tracker Plugin — New Features, Redesigned UI & Fixes

AhDoozy opened the pull request:
This PR delivers a rebuilt Goal Tracker plugin with new functionality, modernized UI, and critical fixes for smoother use.
This rebuild makes Goal Tracker faster, cleaner, and more reliable — introducing a Quest Requirements system, presets, improved visuals, and much smoother task/goal management.

✨ New Features
	•	Shift+Click task removal for faster management
	•	Automatic goal status checks for real-time progress
	•	ActionBar & ActionBarButton UI components for navigation and bulk actions
	•	Quest Requirements file built and integrated for quick access to quest pre-reqs
	•	Preset Goal Lists
	•	Export/Import goals to JSON with full UI refresh
	•	Inline editing for goal titles and manual tasks
	•	Long titles now ellipsized with hover-to-view full text
	•	Export/Import buttons

♻️ Redesigned
	•	Quest prerequisites button leveraging the new Quest Requirements file
	•	Cascading completion to auto-complete related tasks
	•	Dropdown quest selector for faster task addition
	•	Right-click menus streamlined for prerequisites and child task options
	•	Goal/task cards with lighter fills, shadows, hover/press effects, and clear header dividers

🐛 Fixes
	•	Undo/Redo functionality cleanup
	•	ActionBarButton painting glitches resolved
	•	Proper refresh handling in GoalTrackerPanel and ListPanel
	•	Fixed blank/refresh issues when switching panels or logging in
	•	Item icons preload correctly at startup/login/import
	•	Home panel refreshes immediately after task completion
	•	Layout fixes around Export button to avoid overlap
	•	Recursive refresh of child tasks ensures status consistency

 Should fix issues #39 #33 #30 #31

AhDoozy committed:
added marking preset-added goals as complete/incomplete
added color selector for goal completion message.
added check upon goal set and login for completed goal.

AhDoozy committed:
Implement manual toggle for preset tasks, configurable completion color, and login goal evaluation

- Introduced right-click context menu handling in GoalPanel to allow toggling Task.Status for tasks created via presets
- Updated Task model and TaskItemContent to support manual status toggling independent of task source
- Added @Alpha color picker to GoalTrackerConfig for customizing chat message color on task completion (default: FF16ABE5)
- Applied selected color in plugin logic where task completion messages are sent via chat
- Hooked into GameState.LOGGED_IN to invoke goalManager.evaluateAllGoals(), checking for and marking completed tasks based on current game state
- Amended README changelog to clarify new features vs previous "fix" label

AhDoozy committed:
- Implement right-click context menu on parent goals with options to mark all child tasks as completed or incomplete
- Add recursive task status update method to Goal model for batch completion toggling
- Ensure goalTrackerPanel refreshes after task updates triggered by quest chat messages
- Add quest task validation to onGameTick to auto-update completion status on login
- Fix issue where quest task completions were not visually reflected until user re-entered the goal view

AhDoozy committed:
working presets

AhDoozy committed:
working presets

AhDoozy committed:
added making complete/incomplete for all indented as well.

AhDoozy committed:
pre-reqs are added directly under the related item

AhDoozy committed:
pre-reqs are added directly under the related item

AhDoozy committed:
pre-reqs are added directly under the related item

AhDoozy committed:
pre-reqs are added directly under the related item

AhDoozy committed:
pre-reqs are added directly under the related item

AhDoozy committed:
pre-reqs are added directly under the related item

AhDoozy committed:
pre-reqs are added directly under the related item

AhDoozy committed:
borked

AhDoozy committed:
Merge pull request #1 from AhDoozy/Quest-Pre-Recs-windows

borked

AhDoozy committed:
working lovely.

AhDoozy committed:
working lovely.

AhDoozy committed:
Merge pull request #2 from AhDoozy/Quest-Pre-Recs

Quest pre recs

AhDoozy committed:
- Quest dropdown now uses **RuneScape UF** font at a normal crisp size for improved readability.
- ComboBox font scaling updated to use integer point sizes, preventing fuzzy text.
- QuestTaskInput updated to rely on shared ComboBox styling for consistency.

AhDoozy committed:
- Back and Undo buttons added to a new top control bar in goal view; later removed and replaced with a cleaner single bar design.
- Embedded red "< Back" button removed from GoalPanel header, leaving only the goal name input aligned cleanly.
- Goal name input updated to support copy, paste, cut, and select-all actions via both context menu and keyboard shortcuts.
- Remove menu option enhanced to also delete all indented child tasks when removing a parent.
- Remove menu label updated so the "(Shift+Left Click)" hint displays smaller and in gray.

AhDoozy committed:
- Back and Undo buttons added to a new top control bar in goal view; later removed and replaced with a cleaner single bar design.
- Embedded red "< Back" button removed from GoalPanel header, leaving only the goal name input aligned cleanly.
- Goal name input updated to support copy, paste, cut, and select-all actions via both context menu and keyboard shortcuts.
- Remove menu option enhanced to also delete all indented child tasks when removing a parent.
- Remove menu label updated so the "(Shift+Left Click)" hint displays smaller and in gray.

AhDoozy committed:
fixed text button issue

AhDoozy committed:
- Home view updated: "Goal Tracker" title moved to its own header, with a new action bar beneath it containing **+ Add goal**, **Move**, and **Bulk Edit** buttons (the latter two are placeholders marked "Coming soon").

AhDoozy committed:
Merge pull request #3 from AhDoozy/quest-pre-reqs-fixes

Quest pre reqs fixes

AhDoozy committed:
added goal panel action bar but right click is now broken. saving for later.

AhDoozy committed:
it's working again!!!

AhDoozy committed:
Lots O Changes

AhDoozy committed:
copy\paste working.

AhDoozy committed:
copy\paste working.

AhDoozy committed:
updated changelog and readme.

AhDoozy committed:
updated readme, license and home panel header.

AhDoozy committed:
lots of UI updates

AhDoozy committed:
lots of UI updates

AhDoozy committed:
appears everythings working

AhDoozy committed:
appears everythings working

AhDoozy committed:
appears everythings working

AhDoozy committed:
Merge pull request #4 from AhDoozy/Home-panel-redesign

Home panel redesign

AhDoozy committed:
weird overlapping on undo is gone.

AhDoozy committed:
new action bar completed.

AhDoozy committed:
fixed actionbar layout, refresing of icons upon load in.

AhDoozy committed:
Merge pull request #5 from AhDoozy/action-buttons-afterfix

Action buttons afterfix

AhDoozy committed:
updates

AhDoozy committed:
massive improvements.

AhDoozy committed:
massive improvements.

AhDoozy committed:
Pinning working, but right click is odd.

AhDoozy committed:
Pinning working, also added pre-reqs button and revamp presets.

AhDoozy committed:
Merge pull request #6 from AhDoozy/Pinning-new

Pinning new

AhDoozy committed:
more presets

AhDoozy committed:
Imported my V2 version into this branch and updated the paths to go back to the OG path and started removing any\all v2. Currently some stuff still lingers but it's running and working.

AhDoozy committed:
Imported my V2 version into this branch and updated the paths to go back to the OG path and started removing any\all v2. Currently some stuff still lingers but it's running and working.

AhDoozy committed:
Imported my V2 version into this branch and updated the paths to go back to the OG path and started removing any\all v2. Currently some stuff still lingers but it's running and working.

AhDoozy committed:
Imported my V2 version into this branch and updated the paths to go back to the OG path and started removing any\all v2. Currently some stuff still lingers but it's running and working.

AhDoozy committed:
Remove V2 naming and ahdoozy paths

AhDoozy committed:
Merge pull request #7 from AhDoozy/codex/locate-goal-tracker-v2-references

Clean up V2 and AhDoozy references

AhDoozy committed:
Imported my V2 version into this branch and updated the paths to go back to the OG path and started removing any\all v2. Currently some stuff still lingers but it's running and working.

AhDoozy committed:
Merge pull request #8 from AhDoozy/move-back-into-OG

Move back into og

AhDoozy committed:
readme update

DjilanoS commented:
@Toofifty Can you add @AhDoozy as contributor or accept the PR.
The current goal tracker is broken and hasnt been maintained for over a year+.

Toofifty commented on README.md:
```diff
@@ -1,3 +1,56 @@
+
+-----
+<details>
+<summary><span style="margin-left:8px;"><h3 style="display:inline;">📜 Original Goal Tracker v1 Readme & Documentation (created by dillydill123)</h3></span></summary>
```
Happy for you to remove the initial README content

Toofifty commented on README.md:
```diff
@@ -1,3 +1,56 @@
+# 🏆 Goal Tracker
+
+> A reimagining of the Goal Tracker plugin — rebuilt with a more modern UI, powerful new features, and improved stability to help you plan, track, and achieve your Old School RuneScape goals with ease.  
+> It is now faster, cleaner, and more reliable — introducing a Quest Requirements system, presets, improved visuals, and smoother task/goal management.
```
nit: this sounds very AI 😂 and I think a lot is unnecessary for regular users thinking of downloading the plugin. Maybe it could just be a brief overview of what the plugin does instead of how clean the UI is?

AhDoozy replied:
Very Ai indeed 😉 🙃

Toofifty commented on runelite-plugin.properties:
```diff
@@ -1,6 +1,7 @@
-description=Keep track of your goals and complete them automatically
-tags=task,to-do,todo
+author=toofifty
+support=https://github.com/toofifty/goaltracker
```
These should not be changed

```suggestion
author=Toofifty
support=https://github.com/Toofifty/rl-goal-tracker
```

Toofifty commented on src/main/java/com/toofifty/goaltracker/models/task/Task.java:
```diff
@@ -57,6 +61,11 @@ public boolean isNotFullyIndented() {
+    /**
+     * Returns a human-readable name for the task.
+     */
+    public abstract String getDisplayName();
```
I don't think this is used

Toofifty commented on README.md:
```diff
@@ -1,3 +1,56 @@
+✨ New Features
+- Shift+Click task removal for faster management
+- Automatic goal status checks for real-time progress
+- ActionBar & ActionBarButton UI components for navigation and bulk actions
```
Like above comment, this isn't really useful to most users (the changelog you added is great and where this technical stuff should be)

Toofifty reviewed (commented):
Hey @AhDoozy! Thanks heaps for this, I've finally had some time to sit down and try out your changes.

All the new features you added look and work great - I'm not going to scrutinize the code too much since you're much more familiar with it than I am at the moment 😅 though I tried to leave some comments to improve the readme description.

I briefly got it running and did some testing, there's a few issues I think need to be addressed before we can release this:

- Empty state is cut off / does not wrap
<img width="265" height="194" alt="Pasted image 20250827203516" src="https://github.com/user-attachments/assets/8ad3e6ae-c8ca-4802-a069-6bc97fbffe60" />

- Adding pre-reqs adds one too many indents
<img width="235" height="120" alt="Pasted image 20250827203829" src="https://github.com/user-attachments/assets/d998ec54-0a9b-4fe2-98f8-73544c23526f" />

- It's also a bit unclear at a glance what "pre-reqs" includes (especially when there is no quests in the list). Maybe the alert that shows could say something like "There are no quest pre-reqs to add"?
- Mark all as complete/incomplete can't be undone
- The pin/unpin options appear twice in the menu when right-clicking on the progress bar or number, and don't appear when clicking on the border of the goal elements
- Progress bars sometimes stretch too wide
<img width="258" height="230" alt="Pasted image 20250827204208" src="https://github.com/user-attachments/assets/ff103757-4cd4-4515-9801-41b58999f459" />

- Can't re-order goals via the right-click options
- I've gotten into a state where the entire panel crashes when clicking a certain goal (maybe when it contains items?). I think the status might be null. Stacktrace:
```
java.lang.NullPointerException: Cannot invoke "Object.hashCode()" because "pk" is null
	at java.base/java.util.ImmutableCollections$MapN.probe(ImmutableCollections.java:1321)
	at java.base/java.util.ImmutableCollections$MapN.get(ImmutableCollections.java:1235)
	at com.toofifty.goaltracker.ui.TaskItemContent.refresh(TaskItemContent.java:145)
	at com.toofifty.goaltracker.ui.GoalPanel.lambda$new$5(GoalPanel.java:104)
	at com.toofifty.goaltracker.ui.components.ListPanel.buildItemPanel(ListPanel.java:102)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
	at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1625)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:921)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:682)
	at com.toofifty.goaltracker.ui.components.ListPanel.buildItemPanels(ListPanel.java:79)
	at com.toofifty.goaltracker.ui.components.ListPanel.tryBuildList(ListPanel.java:162)
```

Let me know when you've got it into a state you're happy with, and I'll add you as a contributor and we can release this :)

AhDoozy commented:
@Toofifty , thanks for the quick reply and going through that. I'll dive into those last few issues hopefully tonight. 
Also, I had plans for the pre-reqs feature to also support items. _For instance adding a pre-req of 70 attack if there is a task item of a whip._  I've got a hardcoded json with a lot of item requirements, just didn't get around to adding that feature in yet. Had too many ideas and not enough time to polish.

AhDoozy committed:
fixed empty panel text wrapping and progress bar from cutting off. Mark all complete/incomplete is working.

AhDoozy committed:
fixes to goal card sizes and added border when pinned.

AhDoozy committed:
Polish rename, pin/unpin, move, and import/export flows

- Rename box no longer stretches the card and only activates when it should
- Added a right-click Rename option with proper commit/cancel (ESC to cancel)
- Cleaned up Pin/Unpin so there’s only one menu item and unpin refreshes instantly
- Move up/down/top/bottom now actually reorders cards; removed forced A–Z sorting
- Export now asks before overwriting a file
- Import now asks whether to Overwrite, Merge, or Cancel

DjilanoS commented:
@AhDoozy If you have time i'd love to be using this anytime soon!

AhDoozy committed:
- Updated readme
- Fixed issue in .properties

AhDoozy committed:
Merge pull request #9 from AhDoozy/9-1-25-2

9-5-25 PR fixes

AhDoozy commented:
Alright, should be solid now. 

- Goal card typography fixed with reserved progress text width to prevent clipping.
- Home panel placeholder updated to use wrapping text; empty text now displays cleanly without clipping.
- Export to file now prompts before overwriting existing files.
- Import from file now prompts with Overwrite / Merge / Cancel options instead of silently replacing goals.
- Home panel no longer forces alphabetical order (forgot I added this lol); manual Move actions now persist with pinned-first grouping only.

AhDoozy commented:
@Toofifty How's it lookin?

Ranrok commented:
@Toofifty I would love to see this updated if possible

Toofifty commented:
Hey! Code changes LGTM - I'll do another round of QA tonight and get this merged in if all looks good :)

Toofifty reviewed (approved):
Nice work! Give me a bit to submit it to plugin hub. I'll add you as a contributor so you're not blocked by me if you decide to add more features :)

AhDoozy commented:
> Nice work! Give me a bit to submit it to plugin hub. I'll add you as a contributor so you're not blocked by me if you decide to add more features :)

very much appreciated!
